index.js

```diff
@@ -1,29 +1,255 @@
+ *    something, this separate set of caches is used. The resolution of dependencies
+ *    and filepaths is done as usual, but the data is then thrown away at the end of
+ *    the call, as is expected when skipCache is true. The next time hashForDep is
+ *    called with skipCache === true, the temp caches are replaced.
+ * 
+ * Within the PERMANENT and TEMP caches are three actual caches:
+ *  - PACKAGE: the cache of PackageCacheEntry objects with information
+ *    about each package found during the hash-for-dep processing.
+ *
+ *    Having the real path means that when resolving depedencies, we can 
+ *    take the resolved paths, hash them and check quickly for cache entries, 
+ *    speeding creation of the cache. 
+ *
+ *    However, that also means we need a second cache to map the 
+ *    cacheKey(name, dir) value to the hashed realPath.
+ *
+ *  - PATH: the cache of nameDirKey => realPathKey. nameDirKey is a hash of the
+ *    combination of the name and starting directory passed to hashForDep. 
+ *    realPathKey is a hash of the resulting real path to the relevant package.json file.
+ *    See the genRealPathKey() and genNameDirKey() functions.
+ *
+ *  - REAL_FILE_PATH: the cache of a file path to the resolved real path of that file.
+ *    This helps to speed up file path resolution while minimizing the use of costly 
+ *    'fs.statSync' methods.
+ */
+var PERMANENT_CACHES = {
+  PACKAGE: new Cache(),
+  PATH: new Cache(),
+  REAL_FILE_PATH: new Cache(),
+
+  reset: function() {
+    this.PACKAGE = new Cache();
+    this.PATH = new Cache();
+    this.REAL_FILE_PATH = new Cache();
+  }
+}
+
+var TEMP_CACHES = {
+  PACKAGE: new Cache(),
+  PATH: new Cache(),
+  REAL_FILE_PATH: new Cache(),
+
+  reset: function() {
+    this.PACKAGE = new Cache();
+    this.PATH = new Cache();
+    this.REAL_FILE_PATH = new Cache();
+  }
+}
+
+/*
+ * Pointer to the set of caches currently in use. When skipCache is not true it is
+ * set to PERMANENT_CACHES. When it is true, this is set to TEMP_CACHES and the
+ * 'reset' function is called at the end of the hashForDep call.
+ */
+var CACHES;
+
+function setCaches(skipCache) {
+  if (skipCache) {
+    CACHES = TEMP_CACHES;
+    CACHES.reset();
+  } else {
+    CACHES = PERMANENT_CACHES;
+  }
+}
+
+/*
+ * Entries in 'PACKAGE' cache. The idea is to minimize the information needed for
+ * each entry (don't store the entire package.json object), but keep
+ * information around for all parts of computing a hash for the dependency
+ * tree of the original package named by 'name' and 'dir' in hashForDep.
+ * The PackageCacheEntry is created at the time the package.json file is
+ * read in (so we only have to do that once), but some properties
+ * may not be initialized until later (this then supports creating cycles
+ * in the dependencies list, which is supported by Node).
+ */
+var PackageCacheEntry = function(name, version, rootDir, sourceHash) {
```

Comment:
nit-pick for consistency, can we use the following form ?

```js
function PackageCacheEntry(name, version, rootDir, sourceHash) {

}
```

Reply:
Oh, certainly! Not sure why I did it the other way - I don't usually use that form.